src/main/java/pl/edu/icm/pl/mxrdr/extension/workflow/XdsSupplyMissingValuesStep.java

```diff
@@ -0,0 +1,137 @@
+
+    public XdsSupplyMissingValuesStep(Map<String, String> inputParams, DatasetVersionServiceBean versionService) {
+        super(inputParams);
+        this.versionService = versionService;
+    }
+
+    @Override
+    protected WorkflowStepResult.Source runInternal(WorkflowExecutionContext workflowExecutionContext, Path path) throws Exception {
+        List<PatternAndReplacement> actions = prepareReplacementActions(workflowExecutionContext);
+        FileContentReplacer.of(XdsAdjustResultStep.XDS_INP, path)
+                .add(actions)
+                .replace();
+        return Success::new;
+    }
+
+    @Override
+    public WorkflowStepResult resume(WorkflowExecutionContext workflowExecutionContext, Map<String, String> map, String s) {
+        throw new UnsupportedOperationException("This step does not pause");
+    }
+
+    @Override
+    public void rollback(WorkflowExecutionContext workflowExecutionContext, Failure failure) { }
+
+    // -------------------- PRIVATE --------------------
+
+    private List<PatternAndReplacement> prepareReplacementActions(WorkflowExecutionContext workflowExecutionContext) {
+        Map<String, String> replacementValues = readReplacementValues(workflowExecutionContext);
+        return PATTERNS.keySet().stream()
+                .filter(k -> replacementValues.get(k) != null)
+                .map(k -> {
+                    PatternAndReplacement patterns = PATTERNS.get(k);
+                    return FileContentReplacer.PatternAndReplacement.of(patterns.pattern,
+                            String.format(patterns.replacement, replacementValues.get(k)));
+                })
+                .collect(Collectors.toList());
+    }
+
+    private Map<String, String> readReplacementValues(WorkflowExecutionContext executionContext) {
+        return findVersion(executionContext)
+                .map(DatasetVersion::getDatasetFields)
+                .orElse(Collections.emptyList()).stream()
+                .filter(f -> MxrdrMetadataField.DATA_COLLECTION.getValue()
+                        .equals(extractFieldName(f)))
+                .findFirst()
+                .map(DatasetField::getDatasetFieldsChildren)
+                .orElse(Collections.emptyList()).stream()
+                .filter(f -> SOURCE_FIELDS_SET.contains(extractFieldName(f)))
+                .collect(HashMap::new,
+                        (m, f) -> m.put(extractFieldName(f), f.getValue()),
+                        HashMap::putAll);
+    }
+
+    private String extractFieldName(DatasetField field) {
+        return field.getDatasetFieldType().getName();
+    }
+
+    private Optional<DatasetVersion> findVersion(WorkflowExecutionContext executionContext) {
+        WorkflowExecution execution = executionContext.getExecution();
+        DatasetVersion version = versionService.findByVersionNumber(execution.getDatasetId(),
+                execution.getMajorVersionNumber(), execution.getMinorVersionNumber());
```

Comment:
`XdsOutputImportStep` uses `executionContext.getDataset().getEditVersion()` to get access to `DatasetVersion` object. Not sure which one is the proper one, but we should use the same way across the whole process.

Reply:
I think that `executionContext.getDataset().getEditVersion()` is not the way to go (at least in my case) as it returns the latest version when that version is a draft, otherwise it creates a new version.

Reply:
I'm not saying that it is, I'm just noticing the difference. All steps should refer to the same `DatasetVersion` to read and write metadata and should do it in the same way. If they do it differently, or of them is most likely a bug or at least a code smell. Also no step should ever create a new version of the Dataset. Please unify access to the `DatasetVersion` across all steps to remove duplication of the same logic with different implementation in different places. Even if it means changing the other step and is not exactly in scope of the task, regularly doing small improvements wherever you can spot them matters the most in the end for clean and maintainable code.